src/webgpu/api/validation/createRenderPipeline.spec.ts

```diff
@@ -248,6 +249,56 @@ g.test('color_formats_must_be_renderable')
     t.doCreateRenderPipelineTest(isAsync, info.renderable && info.color, descriptor);
   });
 
+g.test('formats_for_depth_stencil_state_must_be_depth_stencil_formats')
+  .params(u => u.combine('isAsync', [false, true]).combine('format', kTextureFormats))
+  .fn(async t => {
+    const { isAsync, format } = t.params;
+    const info = kTextureFormatInfo[format];
+    await t.selectDeviceOrSkipTestCase(info.feature);
+
+    const descriptor = t.getDescriptor({ depthStencil: { format } });
+
+    t.doCreateRenderPipelineTest(isAsync, info.depth || info.stencil, descriptor);
+  });
+
+g.test('format_aspect_requirement_for_depth_stencil_test')
+  .desc(
+    `Tests the format aspect requirement:
+  - Depth aspect must be contained in the format if depth test or depth write is enabled.
+  - Stencil aspect must be contained in the format if stencil test or stencil write is enabled.
+  `
+  )
+  .params(u => u.combine('isAsync', [false, true]).combine('format', kDepthStencilFormats))
+  .fn(async t => {
+    const { isAsync, format } = t.params;
+    const info = kTextureFormatInfo[format];
+    await t.selectDeviceOrSkipTestCase(info.feature);
+
+    // Depth test is enabled
+    const descriptorDepthTest = t.getDescriptor({
+      depthStencil: { format, depthCompare: 'less-equal' },
+    });
+    t.doCreateRenderPipelineTest(isAsync, info.depth, descriptorDepthTest);
+
+    // Depth write is enabled
+    const descriptorDepthWrite = t.getDescriptor({
+      depthStencil: { format, depthWriteEnabled: true },
+    });
+    t.doCreateRenderPipelineTest(isAsync, info.depth, descriptorDepthWrite);
+
+    // Stencil test is enabled
+    const descriptorStencilTest = t.getDescriptor({
+      depthStencil: { format, stencilFront: { compare: 'less-equal' } },
+    });
+    t.doCreateRenderPipelineTest(isAsync, info.stencil, descriptorStencilTest);
+
+    // Stencil write is enabled
+    const descriptorStencilWrite = t.getDescriptor({
+      depthStencil: { format, stencilBack: { failOp: 'replace' } },
+    });
+    t.doCreateRenderPipelineTest(isAsync, info.stencil, descriptorStencilWrite);
```

Comment:
IMO we should have cases that test changing one of every stencil option:
stencilFront.compare
stencilFront.failOp
stencilFront.depthFailOp
stencilFront.passOp
stencilBack.compare
stencilBack.failOp
stencilBack.depthFailOp
stencilBack.passOp

Maybe this is one test parameterized over `stencilFaceKey: 'stencilFront' | 'stencilBack'` and `compare: GPUCompareFunction | undefined`
which does `descriptor.depthStencil[stencilFaceKey].compare = compare`

and another test similarly parameterized over `stencilFaceKey: 'stencilFront' | 'stencilBack'`, `stencilOpKey: 'failOp' | 'depthFailOp' | 'passOp'`, and `op: GPUStencilOperation`.

Reply:
Done. I followed this suggestion but did it differently. Because `descriptor.depthStencil[stencilFaceKey].compare = compare` doesn't work (because stencilFaceKey is a string? and it failed to build via `npx grunt pre` ).

Reply:
To make my suggestion work you would use `as const` to make a string `'stencilFront`' have type `'stencilFront'` instead of type `string`, e.g. `['stencilFront', 'stencilBack'] as const`.
followed I think by `const descriptor = t.getDescriptor({ depthStencil: { format, [stencilFaceKey]: { compare } } });`

However the code you wrote looks just fine as well, and probably easier to read, at least for `stencil_test`.

Reply:
I think it's more readable in `stencil_write` too.

Reply:
Understood. Thanks for the explanation!